### PR TITLE
Implementation for odh uploader to use email from the generated netex

### PR DESCRIPTION
**Pull request by tejpowar**
# Description

Add the ability to email the generated netex

# Testing instructions

-  ODH Uploader has the ability to get the email, uuid and other keys to add into the email. There is a an email template.

In order to test if you checkout the https://github.com/fares-data-build-tool/fdbt-dev/pull/12 and this pr. Under the fdbt-dev branch if you run `make generate-email-odh-uploader`

This will generate the periodMultiService

You can change this to see by changing the filename for the make generator command for example for single you can use

```
generate-email-odh-uploader:
	./scripts/trigger_odh_uploader.sh single
```

This will allow you to test the N/A for products and services as the array does not exist.

No email can be sent locally so you can console.log(mailOptions) from odh-handler
# Type of change

-   [x] feat - A new feature
-   [ ] fix - A bug fix
-   [ ] docs - Documentation only changes
-   [ ] codestyle - Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
-   [ ] refactor - A code change that neither fixes a bug nor adds a feature
-   [ ] perf - A code change that improves performance
-   [ ] test - Adding missing tests or correcting existing tests
-   [ ] build - Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
-   [ ] ci - Changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs)
-   [ ] chore - Other changes that don't modify src or test files
-   [ ] revert - Reverts a previous commit
-   [ ] content - Changes to content or copy

# Checklist:

-   [x] Able to run pr locally
-   [x] Followed acceptance criteria
-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my own code
-   [ ] I have made corresponding changes to the documentation if applicable

### CODE REVIEW

File: src/odh-uploader/handler.test.ts [hunk 1]
```diff
@@ -1,27 +1,22 @@
...
-        mockS3GetObject.mockImplementation(() => ({
-            promise(): Promise<{}> {
-                return Promise.resolve({ Body: testData.testNetexFromS3 });
-            },
-        }));
+    beforeEach(() => {
+        mockFetchDataFromS3Spy.mockImplementation(() =>
+            Promise.resolve({ Body: testData.testNetexFromS3, email: 'test@example.com' }),
+        );
+        mockGetNetexFileFromS3.mockImplementation(() => Promise.resolve('Body: testData.testNetexFromS3'));
```
Comment (dannydaviesIW) [inline src/odh-uploader/handler.test.ts:19]: I think we need to update these mock Promise.resolve() objects. The mockFetchDataFromS3Spy will return a matching data object as far as I can tell? And the mockGetNetexFileFromS3Spy will return a just a string of netex?

File: src/odh-uploader/handler.ts [hunk 2]
```diff
@@ -42,20 +68,28 @@ export const setMailOptions = (params: S3ObjectParameters, pathToNetex: string):
...
-    return data.Body?.toString('utf-8') ?? '';
-};
-
 export const odhUploaderHandler = async (event: S3Event): Promise<void> => {
     try {
         const s3ObjectParams = setS3ObjectParams(event);
         const pathToSavedNetex = `/tmp/${s3ObjectParams.Key}`;
         const netexFile = await getNetexFileFromS3(s3ObjectParams);
+
+        const dataAsString = await fetchDataFromMatchingS3(event);
```
[the remark this answers is not in the document]
Comment (tejpowar) [reply src/odh-uploader/handler.ts:77]: Potentially i can create a matchingData

File: src/odh-uploader/data/s3.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+};
+
+const s3 = getS3Client();
+
+export const getNetexFileFromS3 = async (params: S3ObjectParameters): Promise<string> => {
+    const data = await s3.getObject(params).promise();
+    return data.Body?.toString('utf-8') ?? '';
+};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
```
[the remark this answers is not in the document]
Comment (tejpowar) [reply src/odh-uploader/data/s3.ts:31]: made this a generic type

File: src/odh-uploader/data/s3.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+                        Bucket: s3MatchingBucket,
+                        Key: s3FileName,
+                    })
+                    .promise()
+            ).Body?.toString('utf-8') ?? '';
+        return JSON.parse(dataAsString);
+    } catch (err) {
+        throw new Error(`Error in retrieving data. ${err.stack}`);
+    }
+};
```
[the remark this answers is not in the document]
Comment (tejpowar) [reply src/odh-uploader/data/s3.ts:49]: yes so have added it into utils/s3. However would be better as a separate ticket to refactor the convertor as we would be testing two things which is not scope of the ticket

File: src/odh-uploader/template/emailTemplate.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    passengerType: string,
+    dateTimeCreation: string,
+): string => {
+    const productsList =
+        products &&
+        products
+            .map(product => {
+                return product.productName;
+            })
+            .join(',');
```
[the remark this answers is not in the document]
Comment (tejpowar) [reply src/odh-uploader/template/emailTemplate.ts:17]: Good shout as this makes it clearer

File: src/odh-uploader/template/emailTemplate.ts [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+    const serviceList =
+        services &&
+        services
+            .map(service => {
+                return service.lineName;
+            })
+            .join(',');
```
[the remark this answers is not in the document]
Comment (tejpowar) [reply src/odh-uploader/template/emailTemplate.ts:25]: done as above

File: src/odh-uploader/template/emailTemplate.ts [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+                  </p>
+                  <p
+                    style="color: #0b0c0c; font-family: Arial,sans-serif; -webkit-font-smoothing: antialiased; font-weight: 400;"
+                  >
+                    <strong>References:</strong> ${uuid}
+                  </p>
+                  <p
+                    style="color: #0b0c0c; font-family: Arial,sans-serif; -webkit-font-smoothing: antialiased; font-weight: 400;"
+                  >
+                    <strong>Services(s):</strong> ${!serviceList ? 'N/A' : serviceList.toString()}
```
[the remark this answers is not in the document]
Comment (tejpowar) [reply src/odh-uploader/template/emailTemplate.ts:89]: removed due to the refactoring

File: src/odh-uploader/handler.test.ts (hunk 1 above)
Comment (tejpowar) [reply src/odh-uploader/handler.test.ts:18]: sorted using the matching test data

**Pull request merged**